File: app/src/test/java/com/cargopull/executor_driver/interactor/ServerConnectionUseCaseTest.java
package com.cargopull.executor_driver.interactor;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.only;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import com.cargopull.executor_driver.backend.web.AuthorizationException;
import io.reactivex.Flowable;
import io.reactivex.functions.Consumer;
import io.reactivex.plugins.RxJavaPlugins;
import io.reactivex.schedulers.Schedulers;
import io.reactivex.schedulers.TestScheduler;
import io.reactivex.subscribers.TestSubscriber;
import java.util.concurrent.TimeUnit;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;
import org.reactivestreams.Subscription;

@RunWith(MockitoJUnitRunner.class)
public class ServerConnectionUseCaseTest {

  private ServerConnectionUseCase useCase;

  @Mock
  private ServerConnectionGateway gateway;
  @Mock
  private Consumer<Subscription> consumer;
  private TestScheduler testScheduler;

  @Before
  public void setUp() {
    testScheduler = new TestScheduler();
    RxJavaPlugins.setComputationSchedulerHandler(scheduler -> testScheduler);
    RxJavaPlugins.setSingleSchedulerHandler(scheduler -> Schedulers.trampoline());
    when(gateway.openSocket()).thenReturn(Flowable.never());
    useCase = new ServerConnectionUseCaseImpl(gateway);
  }

  /* Проверяем работу с гейтвеем */

  /**
   * Должен запросить у гейтвея открытие сокета.
   */
  @Test
  public void askGatewayToOpenSocket() {
    // Действие:
    useCase.connect().test();

    // Результат:
    verify(gateway, only()).openSocket();
  }

  /**
   * Не должен просить у гейтвея повторных открытий сокета.
   */
  @Test
  public void doNotAskGatewayToOpenSocketAgain() {
    // Действие:
    useCase.connect().test();
    useCase.connect().test();
    useCase.connect().test();

    // Результат:
    verify(gateway, only()).openSocket();
  }

  /**
   * Должен просить у гейтвея повторные открытия сокета после завершений.
   */
  @Test
  public void askGatewayToOpenSocketAgainAfterComplete() {
    // Дано:
    when(gateway.openSocket()).thenReturn(Flowable.empty());

    // Действие:
    useCase.connect().test();
    useCase.connect().test();
    useCase.connect().test();

    // Результат:
    verify(gateway, times(3)).openSocket();
  }

  /**
   * Должен просить у гейтвея повторные открытия сокета после ошибок авторизации.
   */
  @Test
  public void askGatewayToOpenSocketAgainAfterAuthorizationErrors() {
    // Дано:
    when(gateway.openSocket()).thenReturn(Flowable.error(new AuthorizationException()));

    // Действие:
    useCase.connect().test();
    useCase.connect().test();
    useCase.connect().test();

    // Результат:
    verify(gateway, times(3)).openSocket();
  }

  /**
   * Не должен просить у гейтвея повторных открытий сокета после прочих ошибок.
   */
  @Test
  public void doNotAskGatewayToOpenSocketAgainAfterOtherErrors() {
    // Дано:
    when(gateway.openSocket()).thenReturn(Flowable.error(new Exception()));

    // Действие:
    useCase.connect().test();
    useCase.connect().test();
    useCase.connect().test();

    // Результат:
    verify(gateway, only()).openSocket();
  }

  /**
   * Должен повторить попытку открытия сокета.
   *
   * @throws Exception error
   */
  @Test
  public void retryOpenSocket() throws Exception {
    // Дано:
    when(gateway.openSocket()).thenReturn(
        Flowable.<Boolean>error(Exception::new).doOnSubscribe(consumer)
    );

    // Действие:
    useCase.connect().test();
    testScheduler.advanceTimeBy(60, TimeUnit.MINUTES);

    // Результат:
    verify(consumer, times(241)).accept(any());
  }

  /* Проверяем ответы на запрос соединения */

  /**
   * Должен ответить ошибкой авторизации.
   */
  @Test
  public void answerAuthorizationError() {
    // Дано:
    when(gateway.openSocket()).thenReturn(Flowable.error(new AuthorizationException()));

    // Действие:
    TestSubscriber<Boolean> testSubscriber = useCase.connect().test();

    // Результат:
    testSubscriber.assertError(AuthorizationException.class);
    testSubscriber.assertNoValues();
    testSubscriber.assertNotComplete();
  }

  /**
   * Должен игнорировать иные ошибки, и для новых подписчиков тоже.
   */
  @Test
  public void ignoreOtherError() {
    // Дано:
    when(gateway.openSocket()).thenReturn(Flowable.error(new Exception()));

    // Действие:
    TestSubscriber<Boolean> testSubscriber = useCase.connect().test();
    testScheduler.advanceTimeBy(45, TimeUnit.MINUTES);
    TestSubscriber<Boolean> testSubscriber2 = useCase.connect().test();
    testScheduler.advanceTimeBy(45, TimeUnit.MINUTES);

    // Результат:
    testSubscriber.assertNoErrors();
    testSubscriber.assertNoValues();
    testSubscriber.assertNotComplete();
    testSubscriber2.assertNoErrors();
    testSubscriber2.assertNoValues();
    testSubscriber2.assertNotComplete();
  }

  /**
   * Должен ответить состояниями подключения сети. При этом должен поделить результат между подписчиками.
   */
  @Test
  public void answerConnectionStates() {
    // Дано:
    when(gateway.openSocket()).thenReturn(
        Flowable.intervalRange(0, 6, 0, 30, TimeUnit.SECONDS, testScheduler)
            .switchMap(i -> Flowable.just(i % 2 == 0))
            .concatWith(Flowable.never())
    );

    // Действие:
    TestSubscriber<Boolean> testSubscriber = useCase.connect().test();
    testScheduler.advanceTimeBy(89, TimeUnit.SECONDS);
    TestSubscriber<Boolean> testSubscriber2 = useCase.connect().test();
    testScheduler.advanceTimeBy(900, TimeUnit.SECONDS);

    // Результат:
    testSubscriber.assertNotComplete();
    testSubscriber.assertNoErrors();
    testSubscriber.assertValues(true, false, true, false, true, false);
    testSubscriber2.assertNotComplete();
    testSubscriber2.assertNoErrors();
    testSubscriber2.assertValues(false, true, false);
  }

  /**
   * Должен ответить закрытием сокета.
   */
  @Test
  public void answerConnectionClosed() {
    // Дано:
    when(gateway.openSocket()).thenReturn(Flowable.empty());

    // Действие:
    TestSubscriber<Boolean> testSubscriber = useCase.connect().test();
    testScheduler.advanceTimeBy(45, TimeUnit.MINUTES);

    // Результат:
    testSubscriber.assertComplete();
    testSubscriber.assertNoErrors();
    testSubscriber.assertNoValues();
  }
}

File: app/src/test/java/com/cargopull/executor_driver/interactor/ServerConnectionUseCaseTest.kt
package com.cargopull.executor_driver.interactor

import com.cargopull.executor_driver.UseCaseThreadTestRule
import com.cargopull.executor_driver.backend.web.AuthorizationException
import io.reactivex.BackpressureStrategy
import io.reactivex.Flowable
import io.reactivex.FlowableEmitter
import io.reactivex.functions.Action
import io.reactivex.functions.Consumer
import io.reactivex.plugins.RxJavaPlugins
import io.reactivex.schedulers.TestScheduler
import org.junit.Before
import org.junit.ClassRule
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentMatchers.any
import org.mockito.Mock
import org.mockito.Mockito.*
import org.mockito.junit.MockitoJUnitRunner
import org.reactivestreams.Subscription
import java.util.concurrent.TimeUnit

@RunWith(MockitoJUnitRunner::class)
class ServerConnectionUseCaseTest {

    companion object {

        @ClassRule
        @JvmField
        val classRule = UseCaseThreadTestRule()
    }

    private lateinit var useCase: ServerConnectionUseCase

    @Mock
    private lateinit var serverConnectionGateway: ServerConnectionGateway
    @Mock
    private lateinit var networkConnectionGateway: CommonGateway<Boolean>
    @Mock
    private lateinit var serverSubscriptionConsumer: Consumer<Subscription>
    @Mock
    private lateinit var networkSubscriptionConsumer: Consumer<Subscription>
    @Mock
    private lateinit var action: Action
    private lateinit var testScheduler: TestScheduler

    @Before
    fun setUp() {
        testScheduler = TestScheduler()
        RxJavaPlugins.setComputationSchedulerHandler { scheduler -> testScheduler }
        `when`(networkConnectionGateway.data)
                .thenReturn(Flowable.never<Boolean>().doOnSubscribe(networkSubscriptionConsumer))
        `when`(serverConnectionGateway.socketState)
                .thenReturn(Flowable.never<Boolean>().doOnSubscribe(serverSubscriptionConsumer))
        useCase = ServerConnectionUseCaseImpl(serverConnectionGateway, networkConnectionGateway)
    }

    /* Проверяем работу с гейтвеем подключения к сети */

    /**
     * Должен запросить у гейтвея состояние сети. И подписаться несколько раз.
     */
    @Test
    @Throws(Exception::class)
    fun askGatewayToGetNetworkConnectionState() {
        // Action:
        useCase.connect().test().isDisposed
        useCase.connect().test().isDisposed
        useCase.connect().test().isDisposed

        // Effect:
        verify(networkConnectionGateway, only()).data
        verify(networkSubscriptionConsumer, times(3)).accept(any())
        verifyNoMoreInteractions(networkSubscriptionConsumer)
    }

    /* Проверяем работу с гейтвеем подключения к сокету */

    /**
     * Должен запросить у гейтвея открытие сокета. И не подписаться ни разу до получения состояния
     * сети.
     */
    @Test
    @Throws(Exception::class)
    fun doNotAskGatewayToOpenSocket() {
        // Action:
        useCase.connect().test().isDisposed
        useCase.connect().test().isDisposed
        useCase.connect().test().isDisposed

        // Effect:
        verify(serverConnectionGateway, only()).socketState
        verifyNoInteractions(serverSubscriptionConsumer)
    }

    /**
     * Не должен подписываться на открытие сокета, если сети нет.
     */
    @Test
    fun doNotTouchGatewayIfNotConnected() {
        // Given:
        `when`(networkConnectionGateway.data).thenReturn(Flowable.just(false))

        // Action:
        useCase.connect().test().isDisposed

        // Effect:
        verifyNoInteractions(serverSubscriptionConsumer)
    }

    /**
     * Должен запросить у гейтвея открытие сокета.
     */
    @Test
    @Throws(Exception::class)
    fun askGatewayToOpenSocket() {
        // Given:
        `when`(networkConnectionGateway.data).thenReturn(Flowable.just(true))

        // Action:
        useCase.connect().test().isDisposed
        useCase.connect().test().isDisposed
        useCase.connect().test().isDisposed

        // Effect:
        verify(serverConnectionGateway, only()).socketState
        verify(serverSubscriptionConsumer, times(3)).accept(any())
        verifyNoMoreInteractions(serverSubscriptionConsumer)
    }

    /**
     * Должен запросить у гейтвея закрытие и открытие сокета.
     */
    @Test
    @Throws(Exception::class)
    fun doNotAskGatewayToCloseAndOpenSocketAgain() {
        // Given:
        val inOrder = inOrder(action, serverSubscriptionConsumer)
        `when`(networkConnectionGateway.data).thenReturn(Flowable.just(true, false, true))
        `when`(serverConnectionGateway.socketState).thenReturn(
                Flowable.never<Boolean>()
                        .doOnSubscribe(serverSubscriptionConsumer)
                        .doOnCancel(action)
        )

        // Action:
        useCase.connect().test().isDisposed

        // Effect:
        inOrder.verify(serverSubscriptionConsumer).accept(any())
        inOrder.verify(action).run()
        inOrder.verify(serverSubscriptionConsumer).accept(any())
        inOrder.verifyNoMoreInteractions()
        verifyNoMoreInteractions(serverSubscriptionConsumer, action)
    }

    /**
     * Должен просить у гейтвея повторные открытия сокета после ошибок авторизации.
     */
    @Test
    @Throws(Exception::class)
    fun askGatewayToOpenSocketAgainAfterAuthorizationErrors() {
        // Given:
        `when`(networkConnectionGateway.data).thenReturn(Flowable.just(true))
        `when`(serverConnectionGateway.socketState).thenReturn(
                Flowable.error<Boolean>(AuthorizationException())
                        .doOnSubscribe(serverSubscriptionConsumer)
        )

        // Action:
        useCase.connect().test().isDisposed
        useCase.connect().test().isDisposed
        useCase.connect().test().isDisposed

        // Effect:
        verify(serverSubscriptionConsumer, times(3)).accept(any())
        verifyNoMoreInteractions(serverSubscriptionConsumer)
    }

    /**
     * Должен повторить попытку открытия сокета.
     *
     * @throws Exception error
     */
    @Test
    @Throws(Exception::class)
    fun retryOpenSocket() {
        // Given:
        lateinit var emitter: FlowableEmitter<Boolean>
        `when`(networkConnectionGateway.data).thenReturn(Flowable.just(true))
        `when`(serverConnectionGateway.socketState).thenReturn(
                Flowable.create<Boolean>({ e -> emitter = e }, BackpressureStrategy.BUFFER)
                        .doOnSubscribe(serverSubscriptionConsumer)
        )

        // Action:
        useCase.connect().test().isDisposed
        emitter.onError(Exception())
        emitter.onError(Exception())
        testScheduler.advanceTimeBy(4, TimeUnit.MINUTES)
        emitter.onError(Exception())
        emitter.onError(Exception())

        // Effect:
        verify(serverSubscriptionConsumer, times(5)).accept(any())
    }

    /* Проверяем ответы на запрос соединения */

    /**
     * Должен ответить ошибкой авторизации.
     */
    @Test
    fun answerAuthorizationError() {
        // Given:
        `when`(networkConnectionGateway.data)
                .thenReturn(Flowable.just(true).concatWith(Flowable.never()))
        `when`(serverConnectionGateway.socketState)
                .thenReturn(Flowable.error(AuthorizationException()))

        // Action:
        val testSubscriber = useCase.connect().test()

        // Effect:
        testSubscriber.assertError(AuthorizationException::class.java)
        testSubscriber.assertNoValues()
        testSubscriber.assertNotComplete()
    }

    /**
     * Должен игнорировать иные ошибки.
     */
    @Test
    fun ignoreOtherError() {
        // Given:
        lateinit var emitter: FlowableEmitter<Boolean>
        `when`(networkConnectionGateway.data)
                .thenReturn(Flowable.just(true).concatWith(Flowable.never()))
        `when`(serverConnectionGateway.socketState).thenReturn(
                Flowable.create({ e -> emitter = e }, BackpressureStrategy.BUFFER)
        )

        // Action:
        val testSubscriber = useCase.connect().test()
        emitter.onError(Exception())
        emitter.onError(Exception())
        testScheduler.advanceTimeBy(45, TimeUnit.MILLISECONDS)
        emitter.onError(Exception())
        emitter.onError(Exception())

        // Effect:
        testSubscriber.assertNoErrors()
        testSubscriber.assertNoValues()
        testSubscriber.assertNotComplete()
    }

    /**
     * Должен ответить состояниями подключения сети.
     */
    @Test
    fun answerConnectionStates() {
        // Given:
        `when`(networkConnectionGateway.data)
                .thenReturn(Flowable.just(true).concatWith(Flowable.never()))
        `when`(serverConnectionGateway.socketState).thenReturn(
                Flowable.intervalRange(0, 6, 0, 30, TimeUnit.SECONDS, testScheduler)
                        .switchMap { i -> Flowable.just(i % 2 == 0L) }
                        .concatWith(Flowable.never())
        )

        // Action:
        val testSubscriber = useCase.connect().test()
        testScheduler.advanceTimeBy(989, TimeUnit.SECONDS)

        // Effect:
        testSubscriber.assertNotComplete()
        testSubscriber.assertNoErrors()
        testSubscriber.assertValues(true, false, true, false, true, false)
    }

    /**
     * Должен ответить закрытием сокета.
     */
    @Test
    fun answerConnectionClosed() {
        // Given:
        `when`(networkConnectionGateway.data).thenReturn(Flowable.just(true))
        `when`(serverConnectionGateway.socketState).thenReturn(Flowable.empty())

        // Action:
        val testSubscriber = useCase.connect().test()
        testScheduler.advanceTimeBy(45, TimeUnit.MINUTES)

        // Effect:
        testSubscriber.assertComplete()
        testSubscriber.assertNoErrors()
        testSubscriber.assertNoValues()
    }
}